packages/asset-swapper/src/quote_consumers/exchange_swap_quote_consumer.ts

```diff
@@ -135,41 +132,38 @@ export class ExchangeSwapQuoteConsumer implements SwapQuoteConsumerBase<Exchange
         if (gasPrice !== undefined) {
             assert.isBigNumber('gasPrice', gasPrice);
         }
-
+        if (ethAmount !== undefined) {
+            assert.isBigNumber('ethAmount', ethAmount);
+        }
         const { orders } = quote;
 
         const finalTakerAddress = await swapQuoteConsumerUtils.getTakerAddressOrThrowAsync(this.provider, opts);
-
-        try {
-            let txHash: string;
-            if (quote.type === MarketOperation.Buy) {
-                const { makerAssetFillAmount } = quote;
-                txHash = await this._contractWrappers.exchange
-                    .marketBuyOrdersNoThrow(orders, makerAssetFillAmount, orders.map(o => o.signature))
-                    .sendTransactionAsync({
-                        from: finalTakerAddress,
-                        gas: gasLimit,
-                        gasPrice,
-                    });
-            } else {
-                const { takerAssetFillAmount } = quote;
-                txHash = await this._contractWrappers.exchange
-                    .marketSellOrdersNoThrow(orders, takerAssetFillAmount, orders.map(o => o.signature))
-                    .sendTransactionAsync({
-                        from: finalTakerAddress,
-                        gas: gasLimit,
-                        gasPrice,
-                    });
-            }
-            return txHash;
-        } catch (err) {
-            if (_.includes(err.message, ContractError.SignatureRequestDenied)) {
-                throw new Error(SwapQuoteConsumerError.SignatureRequestDenied);
-            } else if (_.includes(err.message, ForwarderError.CompleteFillFailed)) {
-                throw new Error(SwapQuoteConsumerError.TransactionValueTooLow);
-            } else {
-                throw err;
-            }
+        const value = ethAmount || quote.worstCaseQuoteInfo.protocolFeeInEthAmount;
+        let txHash: string;
+        if (quote.type === MarketOperation.Buy) {
+            const { makerAssetFillAmount } = quote;
+            txHash = await this._exchangeContract
+                .marketBuyOrdersFillOrKill(orders, makerAssetFillAmount, orders.map(o => o.signature))
+                .sendTransactionAsync({
+                    from: finalTakerAddress,
+                    gas: gasLimit,
+                    gasPrice,
+                    value,
+                });
+        } else {
+            const { takerAssetFillAmount } = quote;
+            txHash = await this._exchangeContract
+                .marketSellOrdersFillOrKill(orders, takerAssetFillAmount, orders.map(o => o.signature))
+                .sendTransactionAsync({
+                    from: finalTakerAddress,
+                    gas: gasLimit,
+                    gasPrice,
+                    value,
+                });
         }
+        // TODO(dorothy-zbornak): Handle signature request denied
```

Comment:
I removed the `try/catch` that was here because it was doing nothing. `ContractError.SignatureRequestDenied` didn't actually get thrown by contract wrappers, and `sendTransactionAsync()` only mines the TX immediately on ganache, so you won't detect a failure on mainnet. 

`ContractError.SignatureRequestDenied` denied is easy enough to detect, but maybe the logic should live in another package. Detecting a `IncompleteFill` will have to be accomplished by doing a `callAsync()` at the same block the TX was mined. Let's add tasks for this.

Reply:
Hmm, yeah that try-catch was residual from `asset-buyer`, I do want to ask though, not very familiar with the nuances of the wrappers, how are errors handled? One of the large feedbacks I received at EthWaterloo was the vague errors thrown, don't want us to make `asset-swapper` more opaque with errors.

Reply:
`SignatureRequestDenied` *was* just running a regex against the thrown error message (by Metamask or w/e). 

If the use `awaitTransactionSuccessAsync()` instead of `sendTransactionAsync()`, it will throw if the TX fails. Then we can use `callAsync()` to replay the TX and get the actual rich revert, which gives us deeper detail.